File: snake.py
import curses, time
import random

UP, DOWN, RIGHT, LEFT, ESC = 1,-1,2,-2,0
HEAD = 0
#Init the screen:
s = curses.initscr()
#Set cursor to invisible (0)
curses.curs_set(0)
#Get max height and width:
sh, sw = s.getmaxyx()
#Create new window in the terminal
w = curses.newwin(sh, sw, 0 , 0)
w.keypad(True)
w.timeout(100)

#addch(row, col, symbol)
def draw(win, obj):
    win.addch(obj.y, obj.x, obj.sym)
def erase(win, obj):
    win.addch(obj.y, obj.x, " ")
class food:
    def __init__(self, screen):
        maxHeight, maxWidth = screen.getmaxyx()
        self.x = random.randint(1, maxWidth-1)
        self.y = random.randint(1, maxHeight-1)
        self.sym = '*'


class snakeNode:
    def __init__(self, x, y, sym):
        self.x = int(x)
        self.y = int(y)
        self.sym = str(sym)

class snake:
    def __init__(self, screen, window):
        self.dir = RIGHT

        self.maxHeight, self.maxWidth = screen.getmaxyx()
        headX = random.randint(1,self.maxWidth-1)
        headY = random.randint(1, self.maxHeight-1)
        
        head = snakeNode(headX, headY, ">")
        self.body = [head]
        self.lenght = len(self.body)
        #self.body.append(head)
        self.window = window
        #display()
    def display(self):
        for i in range(0, len(self.body)):
            draw(self.window, self.body[i])
    def move(self, dir):
        #if (dir != self.dir) and (dir != -self.dir):
        if (dir != -self.dir):
            self.dir = dir
        mod = [0,0]
        headSym = ">"
        if self.dir == UP:
            mod[1] = -1
            headSym = "^"
        elif self.dir == DOWN:
            mod[1] = 1
            headSym = "v"
        elif self.dir == RIGHT:
            mod[0] = 1
            headSym = ">"
        elif self.dir == LEFT:
            mod[0] = -1
            headSym = "<"
        newX = self.body[HEAD].x + mod[0]
        newY = self.body[HEAD].y + mod[1]
        if newX >= self.maxWidth:
            newX = 0 + mod[0]
        elif newX <= 0:
            newX = self.maxWidth + mod[0]

        if newY >= self.maxHeight:
            newY = 0 + mod[1]
        elif newY <= 0:
            newY = self.maxHeight + mod[1]
        
        newHead = snakeNode(newX, newY, headSym)
        self.body[HEAD].sym = "o"
        self.body.insert(HEAD, newHead)
        erase(self.window, self.body.pop())
    def eat(self, food):
        if self.body[HEAD].x == food.x and self.body[HEAD].y == food.y:
            #newNode = snakeNode()
            mod = [0,0]
            headSym = ">"
            if self.dir == UP:
                mod[1] = -1
                headSym = "^"
            elif self.dir == DOWN:
                mod[1] = 1
                headSym = "v"
            elif self.dir == RIGHT:
                mod[0] = 1
                headSym = ">"
            elif self.dir == LEFT:
                mod[0] = -1
                headSym = "<"
            newX = self.body[HEAD].x + mod[0]
            newY = self.body[HEAD].y + mod[1]
            if newX >= self.maxWidth:
                newX = 0 + mod[0]
            elif newX <= 0:
                newX = self.maxWidth + mod[0]

            if newY >= self.maxHeight:
                newY = 0 + mod[1]
            elif newY <= 0:
                newY = self.maxHeight + mod[1]
            self.body[HEAD].sym = "o"
            newNode = snakeNode(newX, newY, headSym)
            self.body.insert(HEAD, newNode)
            self.lenght = len(self.body)
            return True
        return False


fd = food(s)
draw(w, fd)

sn = snake(s, w)
sn.display()

snake_dir = RIGHT
key = curses.KEY_RIGHT
while True:
    sn.move(snake_dir)
    sn.display()
    next_key = w.getch()
    if next_key == -1:
        key = key
    else:
        key = next_key
    if sn.eat(fd) == True:
        fd = food(s)
        draw(w, fd)
    for i in range(1, sn.lenght):
        if sn.body[0].x == sn.body[i].x and sn.body[0].y == sn.body[i].y:
            curses.endwin()
            quit()

    if key == curses.KEY_UP:
        snake_dir = UP
    elif key == curses.KEY_DOWN:
        snake_dir = DOWN
    elif key == curses.KEY_RIGHT:
        snake_dir = RIGHT
    elif key == curses.KEY_LEFT:
        snake_dir = LEFT
    elif key == curses.KEY_BACKSPACE:
        break

curses.endwin()
quit()

File: snakeJS/snake.js
const defaultW = 10;
const defaultH = 10;

class node {
    constructor(x, y, color){
        this._x = x;
        this._y = y;
        this._color = color;
    }
    get x()
    {
        return this._x;
    }
    set x(val)
    {
        this._x = val;
    }
    get y()
    {
        return this._y;
    }
    set y(val)
    {
        this._y = val;
    }
    get color()
    {
        return this._color;
    }
    set color(val)
    {
        this._color = val;
    }
    // move(dx_factor, dy_factor)
    // {
    //     this._dx = dx_factor * defaultW;
    //     this._dy = dy_factor * defaultH;
    // }
    // update()
    // {
    //     this._x = (this._x + this._dx + canvasWidth) % canvasWidth;
    //     this._y = (this._y + this._dy + canvasHeight) % canvasHeight;
    // }
    draw() {
        fill(this._color[0], this._color[1], this._color[2]);
        rect(this._x, this._y, defaultW, defaultH)
    }
}
class Snake
{
    constructor(node){
        this._color = node._color;
        this._body = [node];
        this._len  = 1;
        this._dx = 10;
        this._dy = 0;
    }
    move(dx_factor, dy_factor)
    {
        if(this._len > 1)
        {
            let tmpdx = dx_factor * defaultW;
            let tmpdy = dy_factor * defaultH;
            if(this._dx != -tmpdx)
            {
                this._dx = tmpdx;
            }
            if(this._dy != -tmpdy)
            {
                this._dy = tmpdy;
            }
        }
        else{
            this._dx = dx_factor * defaultW;
            this._dy = dy_factor * defaultH;
        }
    }
    update()
    {
        for(let i = 1; i < this._len ;i++)
        {
            if(this._body[0].x == this._body[i].x &&
                this._body[0].y == this._body[i].y)
            {
                return false;
            }
        }
        var newX = (this._body[0].x + this._dx + canvasWidth) % canvasWidth;
        var newY = (this._body[0].y + this._dy + canvasHeight) % canvasHeight;
        var newNode = new node(newX, newY, this._color);
        this._body.pop();
        this._body.unshift(newNode);
        return true;
    }
    eat(food)
    {
        if(this._body[0].x == food.x && this._body[0].y == food.y)
        {
            var newX = (this._body[0].x + this._dx + canvasWidth) % canvasWidth;
            var newY = (this._body[0].y + this._dy + canvasHeight) % canvasHeight;
            var newNode = new node(newX, newY, this._color);
            // this._body.pop();
            this._body.unshift(newNode);
            this._len++;
            return true;
        }
        return false;
    }
    draw()
    {
        for(let i = 0; i < this._len; i++)
        {
            this._body[i].draw();
        }
    }
}
